optuna/_gp/search_space.py

```diff
@@ -28,46 +27,114 @@
 _threading_lock = threading.Lock()
 
 
-class ScaleType(IntEnum):
+class _ScaleType(IntEnum):
     LINEAR = 0
     LOG = 1
     CATEGORICAL = 2
 
 
-@dataclass(frozen=True)
 class SearchSpace:
-    scale_types: np.ndarray
-    bounds: np.ndarray
-    steps: np.ndarray
-
-    @property
-    def is_categorical(self) -> np.ndarray:
-        return self.scale_types == ScaleType.CATEGORICAL
+    def __init__(
+        self,
+        optuna_search_space: dict[str, BaseDistribution],
+    ) -> None:
+        self._optuna_search_space = optuna_search_space
+        self._scale_types = np.zeros(len(optuna_search_space), dtype=np.int64)
+        self._bounds = np.zeros((len(optuna_search_space), 2), dtype=np.float64)
+        self._steps = np.zeros(len(optuna_search_space), dtype=np.float64)
+        for i, distribution in enumerate(optuna_search_space.values()):
+            if isinstance(distribution, CategoricalDistribution):
+                self._scale_types[i] = _ScaleType.CATEGORICAL
+                self._bounds[i, :] = (0.0, len(distribution.choices))
+                self._steps[i] = 1.0
+            else:
+                assert isinstance(distribution, (FloatDistribution, IntDistribution))
+                self._scale_types[i] = _ScaleType.LOG if distribution.log else _ScaleType.LINEAR
+                self._bounds[i, :] = (distribution.low, distribution.high)
+                self._steps[i] = distribution.step or 0.0
+        self.dim = len(optuna_search_space)
+        # TODO: Make it an index array.
+        self.is_categorical = self._scale_types == _ScaleType.CATEGORICAL
+        # NOTE(nabenabe): MyPy Redefinition for NumPy v2.2.0. (Cast signed int to int)
+        self.discrete_indices = np.flatnonzero(self._steps > 0).astype(int)
+        self.continuous_indices = np.flatnonzero(self._steps == 0.0).astype(int)
+
+    def get_normalized_params(
+        self,
+        trials: list[FrozenTrial],
+    ) -> np.ndarray:
+        values = np.zeros((len(trials), len(self._optuna_search_space)), dtype=np.float64)
```

Comment:
[nit] In Optuna, `dtype=float` is more conventional, but it is a nit picking.
https://numpy.org/doc/stable/user/basics.types.html

For a huge array, `np.empty` is better, but the size of this array would not be that huge for `GPSampler` probably.
```suggestion
        values = np.empty((len(trials), len(self._optuna_search_space)), dtype=float)
```

Reply:
Thank you for the suggestion. I was just following the original implementation in order to minimize the difference from the original one. However, I agree with the points you mentioned, and I wanted to update.
Since the same thing also holds for the contractor, I will also update it if we update here.